pluto: add proxy support for FIPS

jpculp opened the pull request:
**Issue number:**
Related: https://github.com/bottlerocket-os/bottlerocket/issues/1667

**Description of changes:**

Replaces `hyper-proxy` with `hyper-http-proxy`.

Adds support for hyper client creation using a custom HTTP connector in `aws-smithy-experimental`.

**Testing done:**

Launched two instances (one built fips and one not) with the following network configuration:
 
```toml
[settings.network]
https-proxy = "<my-proxy>:8888"
```

Checked proxy server logs and found all traffic from my private IP being tunneled, including the EKS API calls for `DescribeCluster`.

In CloudTrail, I could see that the `DescribeCluster` API call source IP was my proxy server:

```
    "eventTime": "2024-10-23T23:15:56Z",
    "eventSource": "eks.amazonaws.com",
    "eventName": "DescribeCluster",
    "awsRegion": "us-west-2",
    "sourceIPAddress": "<my-proxy>",
    "userAgent": "aws-sdk-rust/1.3.3 os/linux lang/rust/1.81.0",
```

**Testing notes:**

If you want to test this on your own, here are a couple of hints:
- I used `tinyproxy` for this PR, but `squidproxy` will also work as I used it for testing https://github.com/bottlerocket-os/bottlerocket/pull/3439.
- In the `tinyproxy` config, uncomment the `LogFile` line and add your EKS cluster's IP range to `Allow`.
- It makes things a little easier to put the proxy in the same VPC/securitygroups as your EKS nodegroups. Then you can set `settings.network.https-proxy` to your EC2 host's private IP DNS name and proxy port.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

jpculp commented:
Based on https://github.com/bottlerocket-os/bottlerocket-core-kit/pull/173.

jpculp commented:
Rebased and ready for review.

ginglis13 commented on sources/api/apiclient/Cargo.toml:
```diff
@@ -23,7 +23,7 @@ futures.workspace = true
 http.workspace = true
 httparse.workspace = true
-hyper = { workspace = true, features = ["client", "http1", "http2", "tcp"] }
+hyper = { version = "0.14", features = ["client", "http1", "http2", "tcp"] }
```
should we add a comment here as to why we're pinning `hyper` to a different version for apiclient?

arnaldo2792 replied:
I think we want to stay (at least for now) on hyper 0.14 for `apiclient` because we use it for the `apiclient exec` feature and we don't want to migrate to a newer hyper version for the FIPS support.

bcressey commented on sources/api/pluto/src/proxy.rs:
```diff
@@ -90,11 +91,19 @@ where
 
-    let https_connector = HttpsConnectorBuilder::new()
-        .with_native_roots()
+    let mut base_connector = HttpConnector::new_with_resolver(GaiResolver::new());
+    base_connector.enforce_http(false);
```
I think it would be cleaner to put the hyper proxy dependency into the smithy client, rather than duplicating this code or exposing more of its methods:
https://gist.github.com/bcressey/d52538b4f7cb13dcbbdf984a0c0f713e

Apart from that, my only misgiving is where to put the `proxy::setup_proxy` interface so that both `pluto` and `cfsignal` can access it.

bcressey replied:
> Apart from that, my only misgiving is where to put the proxy::setup_proxy interface so that both pluto and cfsignal can access it.

Maybe add this as a separate module to the smithy client fork for now, since it has the right dependencies already.

jpculp committed:
pluto: add proxy support for FIPS

Replaces `hyper-proxy` with `hyper-http-proxy`.

Adds support for hyper client creation using a proxy connector in
`aws-smithy-experimental`.

Signed-off-by: Patrick J.P. Culp <jpculp@amazon.com>

jpculp commented:
Moved proxy logic from `pluto::proxy` to `aws-smithy-experimental::hyper_1_0`.

ginglis13 commented on sources/aws-smithy-experimental/src/hyper_1_0.rs:
```diff
@@ -209,6 +213,60 @@ mod build_connector {
+        // Parse https_proxy as URL to extract out auth information if any
+        let proxy_url = Url::parse(https_proxy).expect("Unable to parse HTTPS proxy as URL");
+
+        if !proxy_url.username().is_empty() || proxy_url.password().is_some() {
```
nit: I was thinking this might read more clearly if it were
```suggestion
        if !(proxy_url.username().is_empty() && proxy_url.password().is_none()) {
```

since the only case where we don't set auth is when both username and password aren't provided

jpculp replied:
This seems fine, but since this has been working as is for a bit I'd like to keep this particular PR to just the necessities.

arnaldo2792 commented on COPYRIGHT:
```diff
@@ -14,12 +14,6 @@ operating system images.
-
-=^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
-
 Contains aws-smithy-experimental from
```
Don't we need to add the list of files we modified?

jpculp commented:
- Added list of modified `.rs` files to `COPYRIGHT`.
- Rebased.
